experimental/list_formatter/src/list_formatter.rs

```diff
@@ -46,86 +39,107 @@ impl ListFormatter {
         Ok(Self { data, width })
     }
 
-    fn format_internal<'c, B>(
-        &'c self,
-        values: &[&str],
-        mut builder: B,
-        append_value: fn(B, &str) -> B,
-        append_literal: fn(B, &str) -> B,
-    ) -> B {
+    fn format_internal<W: FormattedWriteable, S: FormattedWriteableSink>(
+        &self,
+        sink: &mut S,
+        values: &[W],
+    ) -> Result<(), S::Error> {
+        macro_rules! literal {
+            ($lit:ident) => {{
+                sink.push_field(Field("literal"))?;
+                sink.write_str($lit)?;
+                sink.pop_field()
+            }};
+        }
+        macro_rules! value {
+            ($val:expr) => {{
+                sink.push_field(Field("element"))?;
+                $val.fmt_write_to(sink)?;
+                sink.pop_field()
+            }};
+        }
+
         match values.len() {
-            0 => builder,
-            1 => append_value(builder, values[0]),
+            0 => Ok(()),
+            1 => value!(values[0]),
             2 => {
                 // Pair(values[0], values[1]) = pair_before + values[0] + pair_between + values[1] + pair_after
-                let (before, between, after) = self.data.get().pair(self.width).parts(values[1]);
-                builder = append_literal(builder, before);
-                builder = append_value(builder, values[0]);
-                builder = append_literal(builder, between);
-                builder = append_value(builder, values[1]);
-                append_literal(builder, after)
+                let (before, between, after) = self.data.get().pair(self.width).parts(&values[1]);
+                literal!(before)?;
```

Comment:
Suggestion: Instead of macros, you could make something like:

```rust
sink.with_field(Field("literal")).write_str(before);
```

where `FormattedSink::with_field()` takes a mutable reference, calls `push_field`, and returns a short-lived object that implements `FormattedSink` and calls `pop_field` when it is dropped (in the `Drop` impl).

Reply:
I like that idea, I tried something similar for lookahead but it got really messy with traits and lifetimes so I went for the simplest API here for now.

How about I submit this and then focus on the detailed design for `FormattedWriteable` and discuss it with the team. It might be a longer process and I don't want to keep this PR lying around. Any `FormattedWriteable` changes will only affect these two experimental crates anyway.